File: packages/snow_draw_core/lib/draw/store/history_delta.dart
import 'package:collection/collection.dart';
import 'package:meta/meta.dart';

import '../models/draw_state.dart';
import '../models/element_state.dart';
import '../models/global_elements_state.dart';
import '../models/interaction_state.dart';
import '../models/selection_overlay_state.dart';
import '../models/selection_state.dart';
import 'snapshot.dart';

@immutable
class HistoryDelta {
  const HistoryDelta._({
    required this.beforeElements,
    required this.afterElements,
    required this.globalElementsBefore,
    required this.globalElementsAfter,
    required this.orderBefore,
    required this.orderAfter,
    required this.orderChanged,
    required this.selectionBefore,
    required this.selectionAfter,
  });

  factory HistoryDelta.fromSnapshots(
    PersistentSnapshot before,
    PersistentSnapshot after,
  ) {
    final beforeById = before.elementMap;
    final afterById = after.elementMap;

    final beforeElements = <String, ElementState>{};
    final afterElements = <String, ElementState>{};
    _collectChangedElementsByScan(
      beforeById: beforeById,
      afterById: afterById,
      beforeElements: beforeElements,
      afterElements: afterElements,
    );

    final orderBefore = List<String>.unmodifiable(before.order);
    final orderAfter = List<String>.unmodifiable(after.order);
    final orderChanged = !const ListEquality<String>().equals(
      orderBefore,
      orderAfter,
    );

    GlobalElementsState? globalElementsBefore;
    GlobalElementsState? globalElementsAfter;
    if (before.globalElements != after.globalElements) {
      globalElementsBefore = before.globalElements;
      globalElementsAfter = after.globalElements;
    }

    SelectionState? selectionBefore;
    SelectionState? selectionAfter;
    if (before.includeSelection &&
        after.includeSelection &&
        before.selection != after.selection) {
      selectionBefore = _copySelection(before.selection);
      selectionAfter = _copySelection(after.selection);
    }

    return HistoryDelta._(
      beforeElements: Map<String, ElementState>.unmodifiable(beforeElements),
      afterElements: Map<String, ElementState>.unmodifiable(afterElements),
      globalElementsBefore: globalElementsBefore,
      globalElementsAfter: globalElementsAfter,
      orderBefore: orderBefore,
      orderAfter: orderAfter,
      orderChanged: orderChanged,
      selectionBefore: selectionBefore,
      selectionAfter: selectionAfter,
    );
  }
  final Map<String, ElementState> beforeElements;
  final Map<String, ElementState> afterElements;
  final GlobalElementsState? globalElementsBefore;
  final GlobalElementsState? globalElementsAfter;
  final List<String> orderBefore;
  final List<String> orderAfter;
  final bool orderChanged;
  final SelectionState? selectionBefore;
  final SelectionState? selectionAfter;

  bool get selectionChanged => selectionBefore != selectionAfter;

  bool get hasChanges =>
      beforeElements.isNotEmpty ||
      afterElements.isNotEmpty ||
      globalElementsBefore != null ||
      globalElementsAfter != null ||
      orderChanged ||
      selectionChanged;

  DrawState applyBackward(DrawState state) => _apply(state, forward: false);

  DrawState applyForward(DrawState state) => _apply(state, forward: true);

  DrawState _apply(DrawState state, {required bool forward}) {
    final currentElements = state.domain.document.elements;
    final currentById = {
      for (final element in currentElements) element.id: element,
    };

    final targetElements = forward ? afterElements : beforeElements;
    final removedIds = forward ? beforeElements.keys : afterElements.keys;
    final retainedIds = forward ? afterElements : beforeElements;
    final nextById = Map<String, ElementState>.from(currentById);

    for (final id in removedIds) {
      if (!retainedIds.containsKey(id)) {
        nextById.remove(id);
      }
    }
    nextById.addAll(targetElements);

    final targetOrder = forward ? orderAfter : orderBefore;

    final nextElements = <ElementState>[];
    for (final id in targetOrder) {
      final element = nextById[id];
      if (element != null) {
        nextElements.add(element);
      }
    }

    final selection = forward ? selectionAfter : selectionBefore;

    return state.copyWith(
      domain: state.domain.copyWith(
        document: state.domain.document.copyWith(
          elements: nextElements,
          globalElements:
              (forward ? globalElementsAfter : globalElementsBefore) ??
              state.domain.document.globalElements,
        ),
        selection: selection ?? state.domain.selection,
      ),
      application: state.application.copyWith(
        interaction: const IdleState(),
        selectionOverlay: SelectionOverlayState.empty,
      ),
    );
  }
}

void _collectChangedElementsByScan({
  required Map<String, ElementState> beforeById,
  required Map<String, ElementState> afterById,
  required Map<String, ElementState> beforeElements,
  required Map<String, ElementState> afterElements,
}) {
  for (final entry in beforeById.entries) {
    final afterElement = afterById[entry.key];
    if (afterElement == null) {
      beforeElements[entry.key] = entry.value;
      continue;
    }
    if (afterElement != entry.value) {
      beforeElements[entry.key] = entry.value;
      afterElements[entry.key] = afterElement;
    }
  }

  for (final entry in afterById.entries) {
    if (!beforeById.containsKey(entry.key)) {
      afterElements[entry.key] = entry.value;
    }
  }
}

SelectionState _copySelection(SelectionState selection) => SelectionState(
  selectedIds: Set<String>.from(selection.selectedIds),
  selectionVersion: selection.selectionVersion,
);

File: packages/snow_draw_rust_engine/src/draw/store/history_delta.rs
#![allow(dead_code)]

use std::collections::BTreeMap;

use crate::draw::models::application_state::{IdleState, InteractionState, SelectionOverlayState};
use crate::draw::models::draw_state::DrawState;
use crate::draw::models::element_state::ElementState;
use crate::draw::models::global_elements_state::GlobalElementsState;
use crate::draw::models::selection_state::SelectionState;
use crate::draw::store::snapshot::PersistentSnapshot;

/// Immutable delta between two history snapshots.
#[derive(Clone, Debug, PartialEq)]
pub struct HistoryDelta {
    pub before_elements: BTreeMap<String, ElementState>,
    pub after_elements: BTreeMap<String, ElementState>,
    pub global_elements_before: Option<GlobalElementsState>,
    pub global_elements_after: Option<GlobalElementsState>,
    pub order_before: Vec<String>,
    pub order_after: Vec<String>,
    pub order_changed: bool,
    pub selection_before: Option<SelectionState>,
    pub selection_after: Option<SelectionState>,
}

impl HistoryDelta {
    /// Creates a history delta from two snapshots.
    pub fn from_snapshots(before: &PersistentSnapshot, after: &PersistentSnapshot) -> Self {
        let mut before_elements = BTreeMap::<String, ElementState>::new();
        let mut after_elements = BTreeMap::<String, ElementState>::new();
        collect_changed_elements_by_scan(
            &before.element_map,
            &after.element_map,
            &mut before_elements,
            &mut after_elements,
        );

        let order_before = before.order.clone();
        let order_after = after.order.clone();
        let order_changed = order_before != order_after;

        let (global_elements_before, global_elements_after) =
            if before.global_elements != after.global_elements {
                (
                    Some(before.global_elements.clone()),
                    Some(after.global_elements.clone()),
                )
            } else {
                (None, None)
            };

        let (selection_before, selection_after) = if before.include_selection
            && after.include_selection
            && before.selection != after.selection
        {
            (
                Some(copy_selection(&before.selection)),
                Some(copy_selection(&after.selection)),
            )
        } else {
            (None, None)
        };

        Self {
            before_elements,
            after_elements,
            global_elements_before,
            global_elements_after,
            order_before,
            order_after,
            order_changed,
            selection_before,
            selection_after,
        }
    }

    /// Whether selection changed in this delta.
    pub fn selection_changed(&self) -> bool {
        self.selection_before != self.selection_after
    }

    /// Whether this delta contains any effective change.
    pub fn has_changes(&self) -> bool {
        !self.before_elements.is_empty()
            || !self.after_elements.is_empty()
            || self.global_elements_before.is_some()
            || self.global_elements_after.is_some()
            || self.order_changed
            || self.selection_changed()
    }

    /// Applies the backward side of the delta to state.
    pub fn apply_backward(&self, state: &DrawState) -> DrawState {
        self.apply(state, false)
    }

    /// Applies the forward side of the delta to state.
    pub fn apply_forward(&self, state: &DrawState) -> DrawState {
        self.apply(state, true)
    }

    fn apply(&self, state: &DrawState, forward: bool) -> DrawState {
        let current_by_id = state
            .domain
            .document
            .elements
            .iter()
            .cloned()
            .map(|element| (element.id.clone(), element))
            .collect::<BTreeMap<_, _>>();

        let target_elements = if forward {
            &self.after_elements
        } else {
            &self.before_elements
        };
        let removed_ids = if forward {
            self.before_elements.keys()
        } else {
            self.after_elements.keys()
        };
        let retained_ids = if forward {
            &self.after_elements
        } else {
            &self.before_elements
        };

        let mut next_by_id = current_by_id;
        for id in removed_ids {
            if !retained_ids.contains_key(id) {
                next_by_id.remove(id);
            }
        }
        for (id, element) in target_elements {
            next_by_id.insert(id.clone(), element.clone());
        }

        let target_order = if forward {
            &self.order_after
        } else {
            &self.order_before
        };
        let mut next_elements = Vec::<ElementState>::with_capacity(target_order.len());
        for id in target_order {
            if let Some(element) = next_by_id.get(id) {
                next_elements.push(element.clone());
            }
        }

        let selection = if forward {
            self.selection_after.as_ref()
        } else {
            self.selection_before.as_ref()
        };
        let global_elements = if forward {
            self.global_elements_after.as_ref()
        } else {
            self.global_elements_before.as_ref()
        };

        let next_document =
            state
                .domain
                .document
                .copy_with(Some(next_elements), global_elements.cloned(), None);
        let next_domain = state.domain.copy_with(
            Some(next_document),
            Some(
                selection
                    .cloned()
                    .unwrap_or_else(|| state.domain.selection.clone()),
            ),
        );
        let next_application = state.application.copy_with(
            None,
            Some(InteractionState::Idle(IdleState)),
            Some(SelectionOverlayState::EMPTY),
        );

        state.copy_with(Some(next_domain), Some(next_application))
    }
}

fn collect_changed_elements_by_scan(
    before_by_id: &BTreeMap<String, ElementState>,
    after_by_id: &BTreeMap<String, ElementState>,
    before_elements: &mut BTreeMap<String, ElementState>,
    after_elements: &mut BTreeMap<String, ElementState>,
) {
    for (id, before_element) in before_by_id {
        match after_by_id.get(id) {
            None => {
                before_elements.insert(id.clone(), before_element.clone());
            }
            Some(after_element) if after_element != before_element => {
                before_elements.insert(id.clone(), before_element.clone());
                after_elements.insert(id.clone(), after_element.clone());
            }
            Some(_) => {}
        }
    }

    for (id, after_element) in after_by_id {
        if !before_by_id.contains_key(id) {
            after_elements.insert(id.clone(), after_element.clone());
        }
    }
}

fn copy_selection(selection: &SelectionState) -> SelectionState {
    SelectionState::new(selection.selected_ids.clone(), selection.selection_version)
}
